File: 01_Array_Hashings/08_Valid_Sudoku/36_Valid_Sudoku.java
import java.util.HashMap;
import java.util.HashSet;

class Solution {
    public boolean isValidSudoku(char[][] board) {
        Map<Integer, Set<Character>> rows =  new HashMap<>();
        Map<Integer, Set<Character>> cols = new HashMap<>();
        Map<String, Set<Character>> squares = new HashMap<>();

        for (int r=0; r<board.length; r++){
            for (int c=0; c<board.length; c++ ){
                System.out.print(board[r][c]);

                if (board[r][c] == '.') { continue;}

                rows.putIfAbsent(r, new HashSet<>());
                cols.putIfAbsent(c, new HashSet<>());
                squares.putIfAbsent((r/3) + "," + (c/3), new HashSet<>());


                if (!rows.get(r).add(board[r][c]) ||
                    !cols.get(c).add(board[r][c]) ||
                    !squares.get((r/3) + "," + (c/3)).add(board[r][c]) ) {
                        return false;
                    }
            }
        }

        return true; 
        
    }
}

File: 01_Array_Hashings/08_Valid_Sudoku/36_Valid_Sudoku.cpp
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {

        unordered_map<int, unordered_set<char>> rows;
        unordered_map<int, unordered_set<char>> cols;
        map<pair<int, int>, unordered_set<char>> squares;

        for (int r=0; r<board.size(); r++){
            for (int c=0; c<board.size(); c++){
                if (board[r][c] == '.'){
                    continue;}

                if (rows[r].count(board[r][c]) != 0 ||
                    cols[c].count(board[r][c]) != 0 ||
                    squares[{r/3, c/3}].count(board[r][c]) != 0){
                        return false;
                }

                rows[r].insert(board[r][c]);
                cols[c].insert(board[r][c]);
                squares[{r/3, c/3}].insert(board[r][c]);

            }
        }

        return true;
        
    }
};